// scikitjs/dist/node/index.js
"use strict";
/**
*  @license
* Copyright 2021, JsData. All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.

* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* ==========================================================================
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecisionTree = exports.DecisionTreeBase = exports.Splitter = exports.RegressionCriterion = exports.ClassificationCriterion = exports.Serialize = exports.fromJSON = exports.fromObject = exports.crossValScore = exports.trainTestSplit = exports.KFold = exports.getBackend = exports.setBackend = exports.makeLowRankMatrix = exports.makeRegression = exports.DecisionTreeRegressor = exports.DecisionTreeClassifier = exports.GaussianNB = exports.LinearSVR = exports.LinearSVC = exports.KNeighborsClassifier = exports.KNeighborsRegressor = exports.VotingClassifier = exports.makeVotingClassifier = exports.VotingRegressor = exports.makeVotingRegressor = exports.dataUrls = exports.KMeans = exports.RobustScaler = exports.ColumnTransformer = exports.makePipeline = exports.Pipeline = exports.Normalizer = exports.OrdinalEncoder = exports.LabelEncoder = exports.OneHotEncoder = exports.SimpleImputer = exports.MaxAbsScaler = exports.StandardScaler = exports.MinMaxScaler = exports.DummyClassifier = exports.DummyRegressor = exports.metrics = exports.LogisticRegression = exports.ElasticNet = exports.RidgeRegression = exports.LassoRegression = exports.LinearRegression = void 0;
var LinearRegression_1 = require("./linear_model/LinearRegression");
Object.defineProperty(exports, "LinearRegression", { enumerable: true, get: function () { return LinearRegression_1.LinearRegression; } });
var LassoRegression_1 = require("./linear_model/LassoRegression");
Object.defineProperty(exports, "LassoRegression", { enumerable: true, get: function () { return LassoRegression_1.LassoRegression; } });
var RidgeRegression_1 = require("./linear_model/RidgeRegression");
Object.defineProperty(exports, "RidgeRegression", { enumerable: true, get: function () { return RidgeRegression_1.RidgeRegression; } });
var ElasticNet_1 = require("./linear_model/ElasticNet");
Object.defineProperty(exports, "ElasticNet", { enumerable: true, get: function () { return ElasticNet_1.ElasticNet; } });
var LogisticRegression_1 = require("./linear_model/LogisticRegression");
Object.defineProperty(exports, "LogisticRegression", { enumerable: true, get: function () { return LogisticRegression_1.LogisticRegression; } });
exports.metrics = __importStar(require("./metrics/metrics"));
var DummyRegressor_1 = require("./dummy/DummyRegressor");
Object.defineProperty(exports, "DummyRegressor", { enumerable: true, get: function () { return DummyRegressor_1.DummyRegressor; } });
var DummyClassifier_1 = require("./dummy/DummyClassifier");
Object.defineProperty(exports, "DummyClassifier", { enumerable: true, get: function () { return DummyClassifier_1.DummyClassifier; } });
var MinMaxScaler_1 = require("./preprocessing/MinMaxScaler");
Object.defineProperty(exports, "MinMaxScaler", { enumerable: true, get: function () { return MinMaxScaler_1.MinMaxScaler; } });
var StandardScaler_1 = require("./preprocessing/StandardScaler");
Object.defineProperty(exports, "StandardScaler", { enumerable: true, get: function () { return StandardScaler_1.StandardScaler; } });
var MaxAbsScaler_1 = require("./preprocessing/MaxAbsScaler");
Object.defineProperty(exports, "MaxAbsScaler", { enumerable: true, get: function () { return MaxAbsScaler_1.MaxAbsScaler; } });
var SimpleImputer_1 = require("./impute/SimpleImputer");
Object.defineProperty(exports, "SimpleImputer", { enumerable: true, get: function () { return SimpleImputer_1.SimpleImputer; } });
var OneHotEncoder_1 = require("./preprocessing/OneHotEncoder");
Object.defineProperty(exports, "OneHotEncoder", { enumerable: true, get: function () { return OneHotEncoder_1.OneHotEncoder; } });
var LabelEncoder_1 = require("./preprocessing/LabelEncoder");
Object.defineProperty(exports, "LabelEncoder", { enumerable: true, get: function () { return LabelEncoder_1.LabelEncoder; } });
var OrdinalEncoder_1 = require("./preprocessing/OrdinalEncoder");
Object.defineProperty(exports, "OrdinalEncoder", { enumerable: true, get: function () { return OrdinalEncoder_1.OrdinalEncoder; } });
var Normalizer_1 = require("./preprocessing/Normalizer");
Object.defineProperty(exports, "Normalizer", { enumerable: true, get: function () { return Normalizer_1.Normalizer; } });
var Pipeline_1 = require("./pipeline/Pipeline");
Object.defineProperty(exports, "Pipeline", { enumerable: true, get: function () { return Pipeline_1.Pipeline; } });
Object.defineProperty(exports, "makePipeline", { enumerable: true, get: function () { return Pipeline_1.makePipeline; } });
var ColumnTransformer_1 = require("./compose/ColumnTransformer");
Object.defineProperty(exports, "ColumnTransformer", { enumerable: true, get: function () { return ColumnTransformer_1.ColumnTransformer; } });
var RobustScaler_1 = require("./preprocessing/RobustScaler");
Object.defineProperty(exports, "RobustScaler", { enumerable: true, get: function () { return RobustScaler_1.RobustScaler; } });
var KMeans_1 = require("./cluster/KMeans");
Object.defineProperty(exports, "KMeans", { enumerable: true, get: function () { return KMeans_1.KMeans; } });
var datasets_1 = require("./datasets/datasets");
Object.defineProperty(exports, "dataUrls", { enumerable: true, get: function () { return datasets_1.dataUrls; } });
var VotingRegressor_1 = require("./ensemble/VotingRegressor");
Object.defineProperty(exports, "makeVotingRegressor", { enumerable: true, get: function () { return VotingRegressor_1.makeVotingRegressor; } });
Object.defineProperty(exports, "VotingRegressor", { enumerable: true, get: function () { return VotingRegressor_1.VotingRegressor; } });
var VotingClassifier_1 = require("./ensemble/VotingClassifier");
Object.defineProperty(exports, "makeVotingClassifier", { enumerable: true, get: function () { return VotingClassifier_1.makeVotingClassifier; } });
Object.defineProperty(exports, "VotingClassifier", { enumerable: true, get: function () { return VotingClassifier_1.VotingClassifier; } });
var KNeighborsRegressor_1 = require("./neighbors/KNeighborsRegressor");
Object.defineProperty(exports, "KNeighborsRegressor", { enumerable: true, get: function () { return KNeighborsRegressor_1.KNeighborsRegressor; } });
var KNeighborsClassifier_1 = require("./neighbors/KNeighborsClassifier");
Object.defineProperty(exports, "KNeighborsClassifier", { enumerable: true, get: function () { return KNeighborsClassifier_1.KNeighborsClassifier; } });
var LinearSVC_1 = require("./svm/LinearSVC");
Object.defineProperty(exports, "LinearSVC", { enumerable: true, get: function () { return LinearSVC_1.LinearSVC; } });
var LinearSVR_1 = require("./svm/LinearSVR");
Object.defineProperty(exports, "LinearSVR", { enumerable: true, get: function () { return LinearSVR_1.LinearSVR; } });
// Comment these out until our libsvm version doesn't ship with fs / path subdependencies
// They were stopping the browser build from being built
// export { SVR, SVRParams } from './svm/SVR'
// export { SVC, SVCParams } from './svm/SVC'
var GaussianNB_1 = require("./naive_bayes/GaussianNB");
Object.defineProperty(exports, "GaussianNB", { enumerable: true, get: function () { return GaussianNB_1.GaussianNB; } });
var DecisionTree_1 = require("./tree/DecisionTree");
Object.defineProperty(exports, "DecisionTreeClassifier", { enumerable: true, get: function () { return DecisionTree_1.DecisionTreeClassifier; } });
Object.defineProperty(exports, "DecisionTreeRegressor", { enumerable: true, get: function () { return DecisionTree_1.DecisionTreeRegressor; } });
var makeRegression_1 = require("./datasets/makeRegression");
Object.defineProperty(exports, "makeRegression", { enumerable: true, get: function () { return makeRegression_1.makeRegression; } });
Object.defineProperty(exports, "makeLowRankMatrix", { enumerable: true, get: function () { return makeRegression_1.makeLowRankMatrix; } });
var tf_singleton_1 = require("./tf-singleton");
Object.defineProperty(exports, "setBackend", { enumerable: true, get: function () { return tf_singleton_1.setBackend; } });
Object.defineProperty(exports, "getBackend", { enumerable: true, get: function () { return tf_singleton_1.getBackend; } });
var KFold_1 = require("./model_selection/KFold");
Object.defineProperty(exports, "KFold", { enumerable: true, get: function () { return KFold_1.KFold; } });
var trainTestSplit_1 = require("./model_selection/trainTestSplit");
Object.defineProperty(exports, "trainTestSplit", { enumerable: true, get: function () { return trainTestSplit_1.trainTestSplit; } });
var crossValScore_1 = require("./model_selection/crossValScore");
Object.defineProperty(exports, "crossValScore", { enumerable: true, get: function () { return crossValScore_1.crossValScore; } });
var simpleSerializer_1 = require("./simpleSerializer");
Object.defineProperty(exports, "fromObject", { enumerable: true, get: function () { return simpleSerializer_1.fromObject; } });
Object.defineProperty(exports, "fromJSON", { enumerable: true, get: function () { return simpleSerializer_1.fromJSON; } });
Object.defineProperty(exports, "Serialize", { enumerable: true, get: function () { return simpleSerializer_1.Serialize; } });
var Criterion_1 = require("./tree/Criterion");
Object.defineProperty(exports, "ClassificationCriterion", { enumerable: true, get: function () { return Criterion_1.ClassificationCriterion; } });
Object.defineProperty(exports, "RegressionCriterion", { enumerable: true, get: function () { return Criterion_1.RegressionCriterion; } });
var Splitter_1 = require("./tree/Splitter");
Object.defineProperty(exports, "Splitter", { enumerable: true, get: function () { return Splitter_1.Splitter; } });
var DecisionTree_2 = require("./tree/DecisionTree");
Object.defineProperty(exports, "DecisionTreeBase", { enumerable: true, get: function () { return DecisionTree_2.DecisionTreeBase; } });
Object.defineProperty(exports, "DecisionTree", { enumerable: true, get: function () { return DecisionTree_2.DecisionTree; } });
